moshi/src/test/java/com/squareup/moshi/JsonAdapterTest.java

```diff
@@ -164,4 +167,71 @@
     serializeNulls.toJson(writer, Collections.<String, String>singletonMap("a", null));
     assertThat(factory.json()).isEqualTo("{\"a\":null}");
   }
+
+  @Test public void defaultIfDataExceptionInt() throws Exception {
+    Moshi moshi = new Moshi.Builder().build();
+    String json = "\"ZERO\"";
+    JsonAdapter<Integer> adapter = moshi.adapter(int.class).defaultIfDataException(0);
+    assertThat(adapter.fromJson(json)).isEqualTo(0);
+  }
+
+  @Test public void defaultIfDataExceptionCustomType() throws Exception {
+    Moshi moshi = new Moshi.Builder().build();
+    String json = "{\"price\":\"eighty\"}";
+    JsonAdapter<Pizza> adapter = moshi.adapter(Pizza.class).defaultIfDataException(new Pizza(-1));
+    assertThat(adapter.fromJson(json)).isEqualTo(new Pizza(-1));
+  }
+
+  @Test public void defaultIfDataExceptionInNested() throws Exception {
+    JsonAdapter.Factory pizzaFactory = new JsonAdapter.Factory() {
+      @Override
+      public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+        if (type != Pizza.class) {
+          return null;
+        }
+        JsonAdapter<Pizza> pizzaAdapter = moshi.nextAdapter(this, type, annotations);
+        return pizzaAdapter.defaultIfDataException(new Pizza(-1));
+      }
+    };
+    Moshi moshi = new Moshi.Builder().add(pizzaFactory).build();
+    String json = "{\"pizza\":{\"price\":\"eighty\"}, \"drink\":\"coke\"}";
+    JsonAdapter<MealDeal> adapter = moshi.adapter(MealDeal.class);
+    assertThat(adapter.fromJson(json)).isEqualTo(new MealDeal(new Pizza(-1), "coke"));
+  }
+
+  private static final class Pizza {
```

Comment:
This feels bad to bring in more lines of machinery than lines of tests. I wonder if TestUtils should contain some test value objects.